[Security Solution] Allow users to edit required_fields field for custom rules

nikitaindik opened the pull request:
**Resolves: https://github.com/elastic/kibana/issues/173594**
**Flaky test runner:** https://buildkite.com/elastic/kibana-flaky-test-suite-runner/builds/5915


## Summary
This PR adds an ability to add and edit custom rule's required fields. "Required fields" is an optional field that shows the user which Elasticsearch fields are needed for the rule to run properly. The values in "required fields" don't affect rule execution in any way. It's purely documentational, similar to "setup guide" and "investigation guide". This functionality is added to both rule creation and rule editing screens. It's available for all rule types except ML.

<img width="650" alt="Scherm­afbeelding 2024-05-07 om 12 28 50" src="https://github.com/elastic/kibana/assets/15949146/70ed052e-3e59-413e-80de-73146225f75a">


## Details
The basic flow goes like this: first you specify your index patterns (or a data view), then you can set required fields for these index patterns. Once a user adds a required field and chooses its name, he can then choose its type from the dropdown. The first available type for the field name selected automatically. User can also add their own custom names and types.

### Warnings
If a field that is not present in the selected index pattern, you will see a warning message.
This can happen in the following cases:
 - You have specified an index pattern, selected a required field from this index pattern, and then removed this index pattern.
 - The index doesn't yet exist. For example, you have installed a prebuilt rule but the data for it hasn't been ingested yet, so there's no index yet.
 - The index was removed.
 - The mappings for the index were changed and the field is no longer present.

In any of these cases, you'll see a general warning message above the form section. And then also a more specific warning message next to the field that is causing the issue.

### ESQL and ML rules
Here's how available dropdown options are determined for different rule types:

For all rule types except ESQL and ML, we take the index patterns specified by the user and fetch their mappings. Then we use these fields and types to populate the dropdowns.

For ESQL rules we parse index patterns out of the query since there's no explicit index pattern form field. We then fetch the mappings for these index patterns and use them to populate the dropdowns.

For ML rules, we don't show "required fields" at all. ML rules are a special case. 
1. The concept of "required fields" is sort of handled during creation of the ML job itself, where the user specifies the fields that are required for the job to run.
2. In the ML rule creation/editing interface, we don't display the index patterns a rule operates on. So, even if we allowed specifying required fields, the user would need to check the ML job details to see the index patterns the job uses.
3. The ML job dropdown includes both existing and not-yet-created jobs. We can't get index patterns for jobs that don't exist yet, so we can't fill the dropdowns with fields and types.

## Screenshots
<img width="628" alt="screen1_" src="https://github.com/elastic/kibana/assets/15949146/aade141f-8285-44c6-8c56-611ba1a9f17b">

<img width="601" alt="screen2_" src="https://github.com/elastic/kibana/assets/15949146/b44fb254-c254-44b8-9600-45b47f29a421">

nikitaindik committed:
Add support for "required_fields" to create and edit rule endpoints

nikitaindik committed:
WIP

banderror commented:
@nikitaindik Can you give an update on the progress on this PR? Is anything blocking you from completing the test coverage and opening it for review?

nikitaindik committed:
WIP

nikitaindik committed:
Fix placeholders and validation

nikitaindik commented:
@banderror Sorry for the late reply. My plan is to wrap up the tests and open the PR for review today.

nikitaindik committed:
Add more tests

nikitaindik commented:
/ci

nikitaindik committed:
Update i18n

nikitaindik committed:
Remove commented out code

nikitaindik commented:
/ci

elasticmachine commented:
Pinging @elastic/security-detections-response (Team:Detections and Resp)

elasticmachine commented:
Pinging @elastic/security-solution (Team: SecuritySolution)

nikitaindik committed:
Update Cypress E2E tests

nikitaindik committed:
Update API integration tests

xcrzx reviewed (commented):
We discussed the UX of required fields with Nikita and Ksenia and decided to introduce changes to make their validation more straightforward from a user perspective: allow adding fields that are not part of the selected index/data view and select their types, show only warnings instead of form errors, and don't block the "Add Required Fields" button. 

I'll continue reviewing the PR once these changes are introduced. But posting a couple of initial comments.

nikitaindik committed:
Remove `related_integrations` from `PrebuiltRuleAsset` type

nikitaindik committed:
Fix UI issues, update tests

nikitaindik committed:
Remove required_fields from baseFields

nikitaindik committed:
Update i18n JSONs

nikitaindik committed:
Merge API integration tests with Related Integerations

nikitaindik committed:
Add a couple API integration tests to check defaulting req. fields value

nikitaindik commented:
/ci

nikitaindik committed:
Fix typos

nikitaindik committed:
Fix a couple integration tests

nikitaindik commented:
/ci

nikitaindik committed:
Update types

nikitaindik commented:
/ci

nikitaindik committed:
Attempt to fix a bug with extracting indices from ESQL queries

nikitaindik commented:
/ci

nikitaindik committed:
Update comments

nikitaindik committed:
Update paddings

nikitaindik commented:
/ci

nikitaindik commented:
/ci

nikitaindik commented:
/ci

nikitaindik commented on x-pack/plugins/security_solution/common/api/detection_engine/model/rule_schema/rule_schemas.schema.yaml:
```diff
@@ -131,6 +131,11 @@ components:
         setup:
           $ref: './common_attributes.schema.yaml#/components/schemas/SetupGuide'
 
+        required_fields:
+          type: array
+          items:
+            $ref: './common_attributes.schema.yaml#/components/schemas/RequiredFieldInput'
+
```
We could do it. It means I'll need to remove `required_fields` from here and add them to `QueryRuleDefaultableFields`, `EqlRuleDefaultableFields`... and so on, right?

Also does it also mean we shouldn't send this field to / from UI for ML rules?

xcrzx replied:
As discussed offline, the `required_fields` field is already a part of the response schema for all rule types. For consistency, let's keep this field in the request schema for all rule types as well.

nikitaindik committed:
Update comments in schemas

nikitaindik committed:
Fix missing "Related Integrations" label in rule creation preview

nikitaindik commented on x-pack/plugins/security_solution/public/detection_engine/rule_creation_ui/components/step_define_rule/schema.tsx:
```diff
@@ -246,6 +246,12 @@ export const schema: FormSchema<DefineStepRule> = {
+    label: i18n.translate(
+      'xpack.securitySolution.detectionEngine.createRule.stepAboutRule.fieldRelatedIntegrationsLabel',
+      {
+        defaultMessage: 'Related integrations',
```
Fixes a missing label in create rule preview
<img width="688" alt="Scherm­afbeelding 2024-05-07 om 12 15 04" src="https://github.com/elastic/kibana/assets/15949146/aaedf303-8189-410a-a462-e1b685692fb4">

xcrzx commented on x-pack/plugins/security_solution/server/lib/detection_engine/prebuilt_rules/model/rule_assets/prebuilt_rule_asset.ts:
```diff
@@ -29,10 +29,15 @@ import {
   z.object({
-    rule_id: RuleSignatureId,
-    version: RuleVersion,
     required_fields: RequiredFieldArray.optional(),
```
Isn't `required_fields` already a part of `BaseCreateProps`?

xcrzx replied:
As discussed offline, we recalculate the `ecs` field when installing rules, so its presence in the rule assets doesn't affect the process. For simplicity, we can omit it.

xcrzx commented on x-pack/plugins/security_solution/public/detection_engine/rule_creation/components/required_fields/required_fields.tsx:
```diff
@@ -0,0 +1,145 @@
+}: RequiredFieldsProps) => {
+  const useFormDataResult = useFormData();
+
+  return (
+    <UseArray path={path} initialNumberOfItems={0}>
+      {({ items, addItem, removeItem, form, error }) => {
+        const [formData] = useFormDataResult;
```
Let's extract this render function to a component for readability.

xcrzx commented on x-pack/plugins/security_solution/common/api/detection_engine/rule_management/import_rules/rule_to_import.ts:
```diff
@@ -29,5 +30,6 @@ export const RuleToImport = BaseCreateProps.and(TypeSpecificCreateProps).and(
   ResponseFields.partial().extend({
     rule_id: RuleSignatureId,
     immutable: z.literal(false).default(false),
+    required_fields: z.array(RequiredFieldInput).optional(),
```
`required_fields` should be already in the structure; no need to redefine it

nikitaindik replied:
Here I am overriding because `ResponseFields` contains `RequiredFieldArray` (with "ecs") and as I understand for import `RequiredFieldInput` (no "ecs") is enough.

nikitaindik committed:
Extract render function into a component, move field logic into hooks, add comments to explain what's going on

nikitaindik committed:
Use `ruleType` instead of `defineStepData.ruleType` (they are same value)

nikitaindik committed:
Fix a typo in comment

xcrzx commented on x-pack/test/security_solution_api_integration/test_suites/detections_response/rules_management/rule_creation/basic_license_essentials_tier/create_rules.ts:
```diff
@@ -207,6 +219,26 @@ export default ({ getService }: FtrProviderContext) => {
       });
+
+      describe('required_fields', () => {
+        afterEach(async () => {
+          await deleteAllRules(supertest, log);
+        });
```
I'm not sure if we need to call `deleteAllRules` for this test case. However, it is generally a good practice to include it in the `beforeEach` hook when necessary. Tests should prepare the database state for themselves before running for stable and predictable results, rather than cleaning it up afterward.

nikitaindik replied:
Actually, I figured it doesn't make sense to call `deleteAllRules` at all, so I removed it. Thanks!

xcrzx commented on x-pack/plugins/security_solution/public/detection_engine/rule_creation_ui/hooks/use_is_valid_esql_query.ts:
```diff
@@ -0,0 +1,62 @@
+ * @param isQueryReadEnabled - if not enabled, return empty array. Useful if we know form or query is not valid and we don't want to retrieve index
+ * @returns boolean - true if field is valid, false if field is invalid or if validation is in progress
+ */
+export const useIsValidEsqlQuery = (
+  query: Query['query'],
+  ruleType: Type,
+  validateFields: FormHook['validateFields']
+) => {
```
As discussed offline, let's try to get rid of this hook altogether.

xcrzx commented on x-pack/test/security_solution_api_integration/test_suites/detections_response/rules_management/rule_update/basic_license_essentials_tier/update_rules.ts:
```diff
@@ -273,6 +279,37 @@ export default ({ getService }: FtrProviderContext) => {
       });
+
+      describe('required_fields', () => {
+        afterEach(async () => {
+          await deleteAllRules(supertest, log);
+        });
```
Same here, it can be removed

nikitaindik committed:
Get rid of the hacky `isValidEsqlQuery` hook

nikitaindik committed:
Update `useEsqlIndex` tests

nikitaindik committed:
Add an explainer comment for `RuleToImport` type

maximpn reviewed (commented):
@nikitaindik thank you for implementing this feature 👍 

I looked through the PR and haven't found any blockers but I think implementation could be simplified. I left comments with my concerns and little code improvements.

I performed quick testing (will perform more through testing later on) and it didn't reveal any blockers as well. I've noticed though ES|QL query shows an error when trying to create an ES|QL rule

![Screenshot 2024-05-08 at 15 50 26](https://github.com/elastic/kibana/assets/3775283/df59dc2b-1122-411a-8eb3-bbe502c252eb)

It should be good to sync offline on my comments and discuss potential improvements.

vitaliidm commented on x-pack/plugins/security_solution/public/detection_engine/rule_creation/components/required_fields/required_fields.tsx:
```diff
@@ -0,0 +1,169 @@
+  path,
+  indexPatternFields = [],
+  isIndexPatternLoading = false,
+}: RequiredFieldsProps) => (
```
Can this component and its children components be memoized to prevent unnecessary re-rendering that may impact page performance?

I've noticed, the more required fields added the more longer it takes for component to re-render

<img width="2546" alt="Screenshot 2024-05-08 at 15 51 08" src="https://github.com/elastic/kibana/assets/92328789/3836b09f-74ed-4479-af10-5a38c67fe999">

Just typing query in search bar, that seemingly does not have impact on required fields causes re-rendering that costs around 5ms, almost 20% of total form render.
With increasing number of fields it maybe even more impactful

nikitaindik replied:
Hey Viltalii! I have investigated the issue thoroughly and ended up fixing it by:
 - wrapping the root `RequiredFields` component into a `React.memo`
 - passing the "watch" array to the `useFormData` hook. Please take a look at the comment and the implementation [here](https://github.com/nikitaindik/kibana/blob/editable-required-fields/x-pack/plugins/security_solution/public/detection_engine/rule_creation/components/required_fields/required_fields.tsx#L80).

With these changes, the `RequiredFields` component no longer re-renders with every query change. Instead, it only re-renders when there are changes to either the "index patterns" or the "required fields".

vitaliidm replied:
good job on tackling the issue

<img width="2556" alt="Screenshot 2024-05-15 at 09 46 31" src="https://github.com/elastic/kibana/assets/92328789/1922c489-2b81-4f2d-a9a2-4bb53a8f6cc1">

vitaliidm commented on x-pack/plugins/security_solution/public/detection_engine/rule_creation/components/required_fields/use_name_field.ts:
```diff
@@ -0,0 +1,126 @@
+  };
+};
+
+function pickTypeForName(
```
can this function be moved to utils and covered with tests?

vitaliidm commented on x-pack/plugins/security_solution/public/detection_engine/rule_creation/components/required_fields/use_type_field.ts:
```diff
@@ -0,0 +1,135 @@
+  handleAddCustomType: (newType: string) => void;
+}
+
+export const useTypeField = (
```
would be great to have description what this and `useNameField` does and what their purpose

nikitaindik replied:
I have moved the contents of this hook into the `TypeComboBox` component and added comments there. I hope it's easier to comprehend now. Please take a look.

nikitaindik commented on x-pack/plugins/security_solution/public/detection_engine/rule_creation/components/required_fields/use_name_field.ts:
```diff
@@ -0,0 +1,126 @@
+  const typesAvailableForNewName = typesByFieldName[currentName] || [];
+  const isCurrentTypeAvailableForNewName = typesAvailableForNewName.includes(currentType);
+
+  let updatedType = currentType;
+
+  /* First try to keep the current type if it's available for the new name */
+  if (isCurrentTypeAvailableForNewName) {
+    return updatedType;
+  }
+
+  /* If it's not available, pick the first available type */
+  if (typesAvailableForNewName.length > 0) {
+    updatedType = typesAvailableForNewName[0];
+  }
+
+  /* Otherwise use currently selected type */
+  return updatedType;
```
Then we'll just return the current type. This should cause a warning to be shown.

Actually, your comment made me take a second look at this function and I figured that I don't really need the `updatedType` variable at all.

nikitaindik replied:
Oh, I just noticed you posted a suggested refactoring above. Thanks!

nikitaindik committed:
Refactor `use_name_field`

vitaliidm commented on x-pack/plugins/security_solution/public/detection_engine/rule_creation/components/required_fields/required_fields_row.tsx:
```diff
@@ -0,0 +1,262 @@
+  };
+  itemId: string;
+}
+
```
<img width="1275" alt="Screenshot 2024-05-08 at 16 18 18" src="https://github.com/elastic/kibana/assets/92328789/b4abb2db-99b6-4a89-bc19-110075055e52">


What's expected behaviour for custom ES|QL created fields? I can see I can add it and then define field type.
But warning still there

nikitaindik replied:
This behavior is actually expected. 

We allow users to add custom fields, even if the index hasn't been created yet. This could be the case, for example, after installing a prebuilt rule. In such scenario, we  to show a warning to indicate that the field doesn't exist, but still allow to save the rule.

So if you add a custom field, you'll see a warning. But once you actually create a mapping for this field, the warning should be gone.

vitaliidm replied:
Mapping won't be created for custom field defined in ES|QL query, but will be saved in alert.
Newly created fields are returned only in ES|QL response, so they always be unmapped, unless user create runtime field.

I think we might want to document this

cc: @nastasha-solomon, @approksiu

nikitaindik committed:
Apply suggested refactoring to `useNameField`

vitaliidm commented on x-pack/plugins/security_solution/public/detection_engine/rule_creation/components/required_fields/required_fields.tsx:
```diff
@@ -0,0 +1,169 @@
+
+  const nameWarnings = fieldValue
+    /* Not creating a warning for empty "name" value */
+    .filter(({ name }) => name !== '')
+    .reduce<Record<string, string>>((warnings, { name }) => {
```
nit: when `reduce` follows `filter`, `filter` logic can be implemented in reduce

nikitaindik committed:
Update `fieldsWithTypes` assignment in `RequiredFieldsList`

vitaliidm reviewed (commented):
I did not reviewed it thoroughly, since @nkhristinin assigned to it.
Just checked ES|QL rule, as its behaviour is different to the rest of rules.
And left note on performance. It might impact page response significantly

nikitaindik committed:
Fix a typo in schema definition comment

nikitaindik commented on x-pack/plugins/security_solution/common/api/detection_engine/model/rule_schema/common_attributes.schema.yaml:
```diff
@@ -315,18 +315,36 @@ components:
+          description: Name of an Elasticsearch field
         type:
           $ref: '../../../model/primitives.schema.yaml#/components/schemas/NonEmptyString'
+          description: Type of the Elasticsearch field
```
No reason really :) Thanks for noticing. Fixed!

nikitaindik committed:
Omit the use of empty `initialState` in `RequiredFields` tests

nikitaindik committed:
Use testIds in form tests

nikitaindik commented on x-pack/plugins/security_solution/public/detection_engine/rule_creation/components/required_fields/required_fields_row.tsx:
```diff
@@ -0,0 +1,262 @@
+    RequiredFieldInput
+  > = useMemo(
+    () => ({
+      type: FIELD_TYPES.JSON,
```
No need indeed. Removed.

nikitaindik committed:
Remove the unnecessary `type` from form field config

nkhristinin commented on x-pack/test/security_solution_cypress/cypress/e2e/detection_response/detection_engine/rule_creation/common_flows.cy.ts:
```diff
@@ -67,6 +68,7 @@ describe('Common rule creation flows', { tags: ['@ess', '@serverless'] }, () =>
   it('Creates and enables a rule', function () {
     cy.log('Filling define section');
     importSavedQuery(this.timelineId);
+    fillRequiredFields();
```
Are we actually check in tests what exaclty we are add to the rule?

Are we want check on Cypress level also warnings?

nikitaindik replied:
> Are we actually check in tests what exaclty we are add to the rule?

Currently we don't. But I agree we should at least check what we send in request. We have discussed this in one of our Tech Sync meetings and decided that this should be done as a part of a greater test refactoring effort.

> Are we want check on Cypress level also warnings?

Since we now try to reduce the number of new Cypress tests we write, I only tested the warnings on the component level by passing different props to it and checking what has rendered. 

I was considered writing a couple of integration-level Jest tests where I would mock part of the form (index patterns and required fields) and then try to add / remove index patterns by typing / clicking. Ultimately I decided that it's not worth the effort since it's tested thoroughly on the component level. wdyt?

vitaliidm replied:
@nikitaindik 
> But I agree we should at least check what we send in request. 

I don't think we need to test what send in request. We just need to assert that these fields are saved and displayed on rule details. That would ensure typed fields are saved correctly within a rule and displayed as expected.

vitaliidm replied:
@nikitaindik , what your thoughts on testing saved fields, displaying in details, instead of asserting request?

nkhristinin commented:
Tested it locally, looks good! You already have a lot of good feedback, but I ask question about cypress tests

dplumlee reviewed (commented):
The implementation is looking good @nikitaindik! I had a small question for you about UI pertaining to the default/empty state:

Before anything is entered, the `helpText` message below the required fields component is present, but the one below related integrations isn't
![Screenshot 2024-05-08 at 3 35 37 PM](https://github.com/elastic/kibana/assets/56367316/aa0aeafa-68d1-470c-beb2-df201c7c6876)

Once the "add" button is clicked for both, they both now show the fields help text and look a lot more consistent.
![Screenshot 2024-05-08 at 3 35 52 PM](https://github.com/elastic/kibana/assets/56367316/7b115209-f86d-4401-86e8-f09d3bb124f1)


I know the help text is displayed under every line for related integrations and it's under the full component itself in required fields so they're a bit different situations but to me it looks a little weird to have the empty component with the help text present as all the other default to having that "empty" state start out displaying an empty input line like these fields (albeit without the help text): 
![Screenshot 2024-05-08 at 3 47 51 PM](https://github.com/elastic/kibana/assets/56367316/66e194ed-fd03-42cf-acef-6f3a523a9b74)

What are your thoughts on making the "empty" state start with one line already present like those forms? Or adding the help text in a similar way to related integrations? I know you mentioned this consistency stuff in a meeting recently so disregard if you're already planning on doing something or deemed it too out of scope for this PR, it just stood out to me when testing this. All these screenshots were taken in the creation form btw, not the editing.

nikitaindik committed:
Update `PrebuiltRuleAsset` type

nikitaindik committed:
Remove unnecessary `RequiredFieldWithOptionalEcs` type

nikitaindik commented:
Thanks everyone for reviewing and leaving feedback! 

Here's what's left for me to do:
 - Figure out the best way to memoize components to avoid re-rendering on every form field change
 - Sync with @ARWNightingale about minor UI tweaks
 - A few refactors, as described in your comments
 - Answer some of your comments

I'll ping everyone once I reopen the PR.

nikitaindik committed:
Update doc link for Related Integrations

nikitaindik committed:
Optimize UI perf by skipping re-renders when irrelevant fields change

nikitaindik committed:
Fix failing API integration test

nikitaindik commented:
/ci

nikitaindik commented:
/ci

nikitaindik committed:
Use a help popover instead of `helpText`

nikitaindik commented:
/ci

nikitaindik committed:
Mock `useKibana` in tests to fix them

nikitaindik commented:
/ci

nikitaindik committed:
Jest tests: Simplify waiting for `handleSubmit` to be called

nikitaindik commented on x-pack/plugins/security_solution/public/detection_engine/rule_creation/components/required_fields/required_fields.test.tsx:
```diff
@@ -0,0 +1,760 @@
+        useForm's "submit" implementation calls setTimeout internally in cases when form is untouched.
+        We need to tell Jest to wait for the next tick of the event loop to allow the form to be submitted.
+      */
+      await waitFor(() => {
```
Thanks! This looks much less hacky.

nikitaindik committed:
Jest tests: Refactor `selectEuiComboBoxOption`

nikitaindik committed:
Jest tests: Use a different testId in warning tests

nikitaindik committed:
Refactor: Use `.some` instead of `.find` to compute `hasEmptyFieldName`

nikitaindik committed:
Refactor: Move `makeValidateRequiredField` into a separate file

nikitaindik committed:
Refactor `pickTypeForName` and cover it with tests

nikitaindik committed:
Refactor: Warning creation

nikitaindik commented on x-pack/plugins/security_solution/server/lib/detection_engine/rule_management/utils/utils.ts:
```diff
@@ -388,3 +391,15 @@ export const migrateLegacyInvestigationFields = (
+
+export const addEcsToRequiredFields = (requiredFields?: RequiredFieldInput[]) =>
+  (requiredFields ?? []).map((requiredFieldWithoutEcs) => {
+    const isEcsField = Boolean(
+      ecsFieldMap[requiredFieldWithoutEcs.name]?.type === requiredFieldWithoutEcs.type
+    );
```
I think we can't. 
`getIsEcsField` returns `true` if field name is in the `ecsFieldMap`, but we also need to check if `type` matches.

nikitaindik committed:
Refactor API integration test: "creates a rule with required_fields defaulted to an empty array when not present"

nikitaindik committed:
Refactor API integration test: "should reset required fields field to default value on update when not present"

nikitaindik committed:
Fix existing bug: display field type icons in Rule Creation and Rule Definition the same way as in Discovery

nikitaindik commented:
/ci

nikitaindik commented:
@maximpn 
> I've noticed though ES|QL query shows an error when trying to create an ES|QL rule

I checked `main` – it behaves the same there. But I agree that seeing an error right away, before touching anything, is weird.

nikitaindik commented on x-pack/plugins/security_solution/public/detection_engine/rule_creation/components/required_fields/required_fields.tsx:
```diff
@@ -0,0 +1,169 @@
+  );
+
+  const allFieldNames = fieldsWithTypes.map(({ name }) => name);
+  const availableFieldNames = allFieldNames.filter((name) => !selectedFieldNames.includes(name));
```
I have investigated this and have concluded that I can't `useMemo` `availableFieldNames` and other related variables because they depend on `fieldValue` and `fieldValue` has a different reference on every render. That's because `useFormData` recreates array field values on every form data change.

But I fixed the issue in a different way by passing the "watch" arg to `useMemo` which makes it re-render only when either "index patterns" or "required fields" themselves change. And if these change, we'll have to compute these values anyways. Feel free to ask questions if it doesn't make sense.

nikitaindik commented:
/ci

nikitaindik committed:
Refactor: wrap selected combobox items into an array just before passing them to `EuiComboBox`

nikitaindik committed:
Commit an automatic tsconfig.json fix

nikitaindik commented:
/ci

nikitaindik commented on x-pack/plugins/security_solution/public/detection_engine/rule_creation/components/required_fields/use_name_field.ts:
```diff
@@ -0,0 +1,126 @@
+      (option) => option.label === field.value.name
+    );
+
+    return selectedNameOption ? [selectedNameOption] : [];
```
No, only a single value is possible. Good idea, I have refactored this piece.

nikitaindik commented:
> What are your thoughts on making the "empty" state start with one line already present like those forms? Or adding the help text in a similar way to related integrations?

Hey @dplumlee! I have discussed this with @ARWNightingale and based on his suggestions I have added the "question mark" popover and moved the description there.

We've also decided to leave the "Field name" / "Field type" placeholders as they are, because otherwise it might be unclear to the user where the field name is and where the type is.

Overall it looks more consistent now.
<img width="648" alt="Scherm­afbeelding 2024-05-14 om 15 47 12" src="https://github.com/elastic/kibana/assets/15949146/806c29d5-0963-4b9a-ac03-a3124d1c27d3">
<img width="650" alt="Scherm­afbeelding 2024-05-14 om 15 47 49" src="https://github.com/elastic/kibana/assets/15949146/f507c680-d5ef-43a5-88f0-da5d74922d79">

nikitaindik commented:
Just noticed that a Jest test fails on CI because I mocked the docs URL incorrectly. Fixing...

nikitaindik committed:
Jest test: Fix incorrectly mocked URL

nikitaindik commented:
Should be fixed now

nikitaindik committed:
Merge branch 'main' into editable-required-fields

vitaliidm commented on x-pack/test/security_solution_cypress/cypress/e2e/detection_response/detection_engine/rule_creation/common_flows.cy.ts:
```diff
@@ -68,7 +70,8 @@ describe.skip('Common rule creation flows', { tags: ['@ess', '@serverless'] }, (
     importSavedQuery(this.timelineId);
-    // The following step is flaky due to a recent EUI upgrade. There is not currently a ticket for the underlying EUI issue, but when there is it will be linked on the test-failure ticket: https://github.com/elastic/kibana/issues/183437
+    // The following steps are flaky due to a recent EUI upgrade. There is not currently a ticket for the underlying EUI issue, but when there is it will be linked on the test-failure ticket: https://github.com/elastic/kibana/issues/183437
+    // fillRequiredFields();
```
the issue refers to failing test on main and due to related integrations. 
Do required fields have the same problem?

vitaliidm commented on x-pack/plugins/security_solution/public/detection_engine/rule_creation/components/required_fields/required_fields.tsx:
```diff
@@ -0,0 +1,211 @@
+    By default, the `useFormData` hook triggers a re-render whenever any form field changes.
+    It also allows optimization by passing a "watch" array of field names. The component then only re-renders when these specified fields change.
+
+    Hovewer, it doesn't work with fields created using the `UseArray` component.
```
```suggestion
    However, it doesn't work with fields created using the `UseArray` component.
```

vitaliidm commented on x-pack/plugins/security_solution/public/detection_engine/rule_creation/components/required_fields/required_fields.tsx:
```diff
@@ -0,0 +1,211 @@
+
+    Hovewer, it doesn't work with fields created using the `UseArray` component.
+    In `useFormData`, these array fields are stored as "flattened" objects with numbered keys, like { "requiredFields[0]": { ... }, "requiredFields[1]": { ... } }.
+    The "watch" feature of `useFormData` only works if you pass these "flattened" field names, such as ["requiredFields[0]", "requiredFields[1]", ...], not just "requiredFields".
```
I wonder, what if you instead of getting all fields, converting it to array with keys and the parsing each key, just reach out directly to `requiredFields__array__` field?

<img width="565" alt="Screenshot 2024-05-15 at 11 29 49" src="https://github.com/elastic/kibana/assets/92328789/166c7788-6b57-4ccb-b19f-ffdd13d012f0">

Looks like it has in value property all the keys you need to watch

nikitaindik committed:
Fix: Add missing icons

nikitaindik committed:
Merge branch 'main' into editable-required-fields

nikitaindik committed:
Refactor: remove unnecessary optional chaining in `pickTypeForName`

nikitaindik commented on x-pack/plugins/security_solution/public/detection_engine/rule_creation/components/required_fields/required_fields_row.tsx:
```diff
@@ -0,0 +1,163 @@
+  itemId: string;
+}
+
+const RequiredFieldField = ({
```
I had a conversation about it with @maximpn and we couldn't figure out a good name.
Maxim's idea was to have `Field` at the end to indicate that it's a component that's passed to `UseField`.
Originally it was called `RequiredFieldRowContent`, which also sounds kind of generic.

Can you suggest a good name for it?

vitaliidm replied:
`RequiredFieldInputs`, `RequiredFieldRowContent` or `RequiredFieldRowItem` 

I think any of them would be preferable

nikitaindik commented on x-pack/plugins/security_solution/public/detection_engine/rule_creation/components/required_fields/required_fields.test.tsx:
```diff
@@ -0,0 +1,734 @@
+      optionText: 'customType',
+    });
+
+    expect(screen.getByDisplayValue('customType')).toBeVisible();
```
I cover warning cases in tests under `describe('warnings'), () => { ... }`. Do you think it's useful to also assert here?

vitaliidm replied:
I did not find the same use case in warnings section (typing custom field) - so, I think it would useful to add it here on in warnings section

nikitaindik committed:
Fix a typo

nikitaindik committed:
Refactor: getting `flattenedFieldNames` in `RequiredFieldsList`

joepeeples commented on x-pack/plugins/security_solution/public/detection_engine/rule_creation/components/required_fields/required_fields_help_info.tsx:
```diff
@@ -0,0 +1,55 @@
+      <EuiText style={{ width: POPOVER_WIDTH }} size="s">
+        <FormattedMessage
+          id="xpack.securitySolution.detectionEngine.ruleDescription.requiredFields.fieldRequiredFieldsHelpText"
+          defaultMessage="Choose the fields that are needed for this rule to execute. You can read more about required fields in our {docLink}."
```
Tightened up the popover text. I also suggest omitting the doc link, since we can probably tell the user all they need in the UI, and the docs won't have much more to say. Also we don't have a very targeted location to send users to: the [current destination](https://www.elastic.co/guide/en/security/master/rules-ui-create.html) is a really long topic, and the "required fields" step will be scattered across the page.
```suggestion
          defaultMessage="Choose the fields and data types needed for this rule to function. You can select the fields available in the rule's **Source** index patterns or data view."
```

joepeeples commented on x-pack/plugins/security_solution/public/detection_engine/rule_creation/components/required_fields/translations.ts:
```diff
@@ -0,0 +1,105 @@
+export const REQUIRED_FIELDS_GENERAL_WARNING_TITLE = i18n.translate(
+  'xpack.securitySolution.detectionEngine.ruleDescription.requiredFields.generalWarningTitle',
+  {
+    defaultMessage: 'Some fields are not found within specified index patterns.',
+  }
+);
+
+export const REQUIRED_FIELDS_GENERAL_WARNING_DESCRIPTION = i18n.translate(
+  'xpack.securitySolution.detectionEngine.ruleDescription.requiredFields.generalWarningDescription',
+  {
+    defaultMessage: `This doesn't break rule execution, but it might indicate that required fields were set incorrectly. Please check that indices specified in index patterns exist and have expected fields and types in mappings.`,
```
Want to make it a little clearer what "specified index patterns" means — specified where?

I don't know if this component allows bold type, but I suggested it to help identify the UI element where the index patterns are specified. Maybe use quotes if bold isn't available.
```suggestion
    defaultMessage: 'Some fields aren't found within the rule's specified index patterns.',
  }
);

export const REQUIRED_FIELDS_GENERAL_WARNING_DESCRIPTION = i18n.translate(
  'xpack.securitySolution.detectionEngine.ruleDescription.requiredFields.generalWarningDescription',
  {
    defaultMessage: `This doesn't break rule execution, but it might indicate that required fields were set incorrectly. Please check that indices specified in the rule's **Source** index patterns exist and have expected fields and types in mappings.`,
```

joepeeples commented on x-pack/plugins/security_solution/public/detection_engine/rule_creation/components/required_fields/translations.ts:
```diff
@@ -0,0 +1,105 @@
+    'xpack.securitySolution.detectionEngine.ruleDescription.requiredFields.fieldNameNotFoundWarning',
+    {
+      values: { name },
+      defaultMessage: `Field "{name}" is not found within specified index patterns`,
```
```suggestion
      defaultMessage: `Field "{name}" is not found within the rule's specified index patterns`,
```

joepeeples commented on x-pack/plugins/security_solution/public/detection_engine/rule_creation/components/required_fields/translations.ts:
```diff
@@ -0,0 +1,105 @@
+    'xpack.securitySolution.detectionEngine.ruleDescription.requiredFields.fieldTypeNotFoundWarning',
+    {
+      values: { name, type },
+      defaultMessage: `Field "{name}" with type "{type}" is not found within specified index patterns`,
```
```suggestion
      defaultMessage: `Field "{name}" with type "{type}" is not found within the rule's specified index patterns`,
```

joepeeples reviewed (commented):
Feature looks great! I suggested some tweaks to UI text here and there, let me know if you have any questions. Thanks!

xcrzx commented on x-pack/plugins/security_solution/public/detection_engine/rule_creation/components/required_fields/required_fields.tsx:
```diff
@@ -0,0 +1,209 @@
+    [indexPatternFields]
+  );
+
+  const allFieldNames = useMemo(() => fieldsWithTypes.map(({ name }) => name), [fieldsWithTypes]);
```
Nit: Do we still need to use `useMemo` in this component? There are a lot of moving parts, and most of the variables in this component are written without using `useMemo`. I'd say we either need to wrap all of them with memoization or none, but not a mix of the two approaches. If the performance is okay, given the `watch` optimization, I'd say we can omit using `useMemo`.

nikitaindik replied:
I thought about removing this, but then decided to leave it in, because `fieldsWithTypes ` array can be really long – almost 2000 items for default index patterns.

xcrzx reviewed (approved):
Tested the PR locally, and everything works as expected. The code changes also look good to me. Thank you for the implementation, @nikitaindik 👍

nikitaindik committed:
Update UI copy

dplumlee reviewed (approved):
Thanks for addressing my comments @nikitaindik, I like the look of the default state a lot more now!

nikitaindik committed:
Merge branch 'main' into editable-required-fields

nikitaindik committed:
Merge branch 'main' into editable-required-fields

nikitaindik committed:
Merge remote-tracking branch 'upstream/main' into editable-required-fields

vitaliidm commented on x-pack/plugins/security_solution/public/detection_engine/rule_creation/components/required_fields/required_fields.tsx:
```diff
@@ -0,0 +1,211 @@
+    The "watch" feature of `useFormData` only works if you pass these "flattened" field names, such as ["requiredFields[0]", "requiredFields[1]", ...], not just "requiredFields".
+
+    To work around this, we manually construct a list of "flattened" field names to watch, based on the current state of the form.
+    This is a temporary solution and ideally, `useFormData` should be updated to handle this scenario.
```
Yes, please, keep me posted or document any findings in the separate issue

jpdjere commented:
### Exploratory testing **In-Progress**

Tested and working as expected:
- Fields appearing as options -with their expected types- when indexing documents into a single **index** that matches the selected index pattern ✅ 
- Fields appearing as options when indexing documents into a multiple indices that match the selected index pattern(s) - **TBD**
- Fields appearing as options when selecting a **Data View** that contains the same index as before ✅ 
- Warning displayed when:
    - You have specified an index pattern, selected a required field from this index pattern, and then removed this index pattern. ✅ _**(warning doesn't appear until another index pattern or data view is selected, but it's acceptable)**_
    - The index doesn't yet exist. For example, you have installed a prebuilt rule but the data for it hasn't been ingested yet, so there's no index yet. ✅ 
    - The index was removed. ✅ 
    - The mappings for the index were changed and the field is no longer present.

joepeeples commented on x-pack/plugins/security_solution/public/detection_engine/rule_creation/components/required_fields/required_fields_help_info.tsx:
```diff
@@ -0,0 +1,48 @@
+      <EuiText style={{ width: POPOVER_WIDTH }} size="s">
+        <FormattedMessage
+          id="xpack.securitySolution.detectionEngine.ruleDescription.requiredFields.fieldRequiredFieldsHelpText"
+          defaultMessage="Choose the fields and data types needed for this rule to function. You can select the fields available in the rule's {source} index patterns or data view."
```
@nikitaindik Here's some revised popover text, after our chat about custom fields. I also wanted to make it clearer that this is just informational — the user isn't actually configuring the rule's logic to _require_ these fields, they're basically just documenting how the rule works.
```suggestion
          defaultMessage="Create an informational list of fields and data types this rule needs to function. Select fields in the rule's {source} index patterns or data view, or type in custom fields."
```

nikitaindik committed:
Merge remote-tracking branch 'upstream/main' into editable-required-fields

nikitaindik committed:
Update tooltip text

nikitaindik committed:
Refactor: remove unnecessary `fieldsWithTypes` computation

nikitaindik committed:
Jest tests: Wrap into `I18nProvider` to remove warnings

nikitaindik committed:
Jest tests: Add a warning assertion

nikitaindik committed:
Related Integrations: Allow to remove the last row to match Required Fields behavior

nikitaindik committed:
Merge remote-tracking branch 'upstream/main' into editable-required-fields
